### ISSUE TITLE
Ubuntu without openmp lead to segment fault

### ISSUE THREAD

**Issue by nwpu2fjl** [intel/mkl-dnn#451]
I built the mkldnn.a without OpenMP(followed the issue #166, I uncomment the line shown below ), but got errors shown below(SegFault) when I run ctest. 

![image](https://user-images.githubusercontent.com/12624852/55713201-bc891880-5a22-11e9-84ca-81281c34e962.png)

![image](https://user-images.githubusercontent.com/12624852/55712716-d0804a80-5a21-11e9-92b7-7a66570f7568.png)

BTW, it is right on macOS. I wonder why it caused SegFault on Ubuntu and how to build MKL-DNN without OpenMP.

**Comment by emfomenk** [comment]
Hi @nwpu2fjl,

I wasn't able to reproduce the behavior you observe. Could you please manually run one of the tests (e.g. simple-net-c) and show how and where exactly it fails? (gdb backtrace would be useful, as well as build with debug `-DCMAKE_BUILD_TYPE=Debug`).

In general, changing the cmake files in any way is not supported model -- there might be some dependencies between the files and if you just comment few of them something might go wrong. If your intention to build the library w/o OpenMP and Intel MKL you can simply pass:
``` bash
  cmake -DMKLDNN_THREADING=SEQ -DMKLDNN_USE_MKL=NONE <other options> ..
```

That should build sequential Intel MKL-DNN.

Out of curiosity, why do you want to use the library w/o OpenMP (or TBB)? Parallelization is quiet essential to get good performance. If you don't want to use external dependencies (like Intel OpenMP that we use by default), you can use:
``` bash
  cmake -DMKLDNN_THREADING=OMP:COMP -DMKLDNN_USE_MKL=NONE <other options>  ..
```

which would make Intel MKL-DNN use compilers OpenMP (GNU OpenMP in case of gcc).

**Comment by nwpu2fjl** [comment]
Thank for replying. There are two reasons why I wanna use MKL-DNN w/o OpenMP. The first is that I wanna compare performance in a single thread, and another is the error caused by importing libasan.so.

**Comment by nwpu2fjl** [comment]
After using the below command and make.
 `cmake -DMKLDNN_THREADING=SEQ -DMKLDNN_USE_MKL=NONE <other options> ..`
I got compiling errors.
[ 50%] Building CXX object tests/gtests/CMakeFiles/test_rnn_forward.dir/main.cpp.o
[ 50%] Building CXX object tests/gtests/CMakeFiles/test_rnn_forward.dir/test_rnn_forward.cpp.o
[ 50%] Linking CXX executable test_rnn_forward
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::UnitTestImpl::GetTestPartResultReporterForCurrentThread()':
gtest-all.cc:(.text+0xd8f4): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text+0xda10): undefined reference to `pthread_setspecific'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::UnitTestImpl::SetTestPartResultReporterForCurrentThread(testing::TestPartResultReporterInterface*)':
gtest-all.cc:(.text+0xf5f9): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text+0xf723): undefined reference to `pthread_setspecific'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::UnitTestImpl::~UnitTestImpl()':
gtest-all.cc:(.text+0xf8ec): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text+0xf907): undefined reference to `pthread_key_delete'
gtest-all.cc:(.text+0xfa0b): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text+0xfa26): undefined reference to `pthread_key_delete'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::UnitTestImpl::UnitTestImpl(testing::UnitTest*)':
gtest-all.cc:(.text+0x11813): undefined reference to `pthread_key_create'
gtest-all.cc:(.text+0x1198b): undefined reference to `pthread_key_create'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::UnitTest::AddTestPartResult(testing::TestPartResult::Type, char const*, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
gtest-all.cc:(.text+0x174b5): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text+0x1751c): undefined reference to `pthread_setspecific'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::ThreadLocal<testing::TestPartResultReporterInterface*>::~ThreadLocal()':
gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED2Ev[_ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED5Ev]+0x20): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED2Ev[_ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED5Ev]+0x35): undefined reference to `pthread_key_delete'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::ThreadLocal<std::vector<testing::internal::TraceInfo, std::allocator<testing::internal::TraceInfo> > >::~ThreadLocal()':
gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED2Ev[_ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED5Ev]+0x20): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED2Ev[_ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED5Ev]+0x35): undefined reference to `pthread_key_delete'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::ThreadLocal<std::vector<testing::internal::TraceInfo, std::allocator<testing::internal::TraceInfo> > >::GetOrCreateValue() const':
gtest-all.cc:(.text._ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv[_ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv]+0x20): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text._ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv[_ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv]+0xa9): undefined reference to `pthread_setspecific'
collect2: error: ld returned 1 exit status
tests/gtests/CMakeFiles/test_rnn_forward.dir/build.make:122: recipe for target 'tests/gtests/test_rnn_forward' failed
make[2]: *** [tests/gtests/test_rnn_forward] Error 1
CMakeFiles/Makefile2:634: recipe for target 'tests/gtests/CMakeFiles/test_rnn_forward.dir/all' failed
make[1]: *** [tests/gtests/CMakeFiles/test_rnn_forward.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
[ 50%] Building CXX object tests/gtests/CMakeFiles/test_rnn_forward.dir/main.cpp.o
[ 50%] Building CXX object tests/gtests/CMakeFiles/test_rnn_forward.dir/test_rnn_forward.cpp.o
[ 50%] Linking CXX executable test_rnn_forward
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::UnitTestImpl::GetTestPartResultReporterForCurrentThread()':
gtest-all.cc:(.text+0xd8f4): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text+0xda10): undefined reference to `pthread_setspecific'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::UnitTestImpl::SetTestPartResultReporterForCurrentThread(testing::TestPartResultReporterInterface*)':
gtest-all.cc:(.text+0xf5f9): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text+0xf723): undefined reference to `pthread_setspecific'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::UnitTestImpl::~UnitTestImpl()':
gtest-all.cc:(.text+0xf8ec): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text+0xf907): undefined reference to `pthread_key_delete'
gtest-all.cc:(.text+0xfa0b): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text+0xfa26): undefined reference to `pthread_key_delete'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::UnitTestImpl::UnitTestImpl(testing::UnitTest*)':
gtest-all.cc:(.text+0x11813): undefined reference to `pthread_key_create'
gtest-all.cc:(.text+0x1198b): undefined reference to `pthread_key_create'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::UnitTest::AddTestPartResult(testing::TestPartResult::Type, char const*, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
gtest-all.cc:(.text+0x174b5): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text+0x1751c): undefined reference to `pthread_setspecific'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::ThreadLocal<testing::TestPartResultReporterInterface*>::~ThreadLocal()':
gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED2Ev[_ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED5Ev]+0x20): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED2Ev[_ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED5Ev]+0x35): undefined reference to `pthread_key_delete'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::ThreadLocal<std::vector<testing::internal::TraceInfo, std::allocator<testing::internal::TraceInfo> > >::~ThreadLocal()':
gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED2Ev[_ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED5Ev]+0x20): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED2Ev[_ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED5Ev]+0x35): undefined reference to `pthread_key_delete'
gtest/libmkldnn_gtest.a(gtest-all.cc.o): In function `testing::internal::ThreadLocal<std::vector<testing::internal::TraceInfo, std::allocator<testing::internal::TraceInfo> > >::GetOrCreateValue() const':
gtest-all.cc:(.text._ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv[_ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv]+0x20): undefined reference to `pthread_getspecific'
gtest-all.cc:(.text._ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv[_ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv]+0xa9): undefined reference to `pthread_setspecific'
collect2: error: ld returned 1 exit status
tests/gtests/CMakeFiles/test_rnn_forward.dir/build.make:122: recipe for target 'tests/gtests/test_rnn_forward' failed
make[2]: *** [tests/gtests/test_rnn_forward] Error 1
CMakeFiles/Makefile2:634: recipe for target 'tests/gtests/CMakeFiles/test_rnn_forward.dir/all' failed
make[1]: *** [tests/gtests/CMakeFiles/test_rnn_forward.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by emfomenk** [comment]
Hi @nwpu2fjl,

Seems like pthread dependency hasn't been propagated for some reasons. However these are gtests only, maybe the library itself is fine. I will take a look tomorrow morning.

What Ubuntu version do you use?

> BTW, How can I know the numbers of threads with OpenMP?

By default the number of threads would most likely be equal to the number of the logical cores in the system (see `lscpu`, `/proc/cpuinfo`, or `numactl -H`).
The `KMP_AFFINITY=verbose` works for Intel OpenMP only. You can check if you use it by calling `ldd app.out` and seeing `libiomp5.so` in the dependencies.

Also, once build with OpenMP you can force the library to use only 1 thread by exporting `OMP_NUM_THREADS=1`.

**Comment by nwpu2fjl** [comment]
But the Segmentation fault hasn't been fixed yet.  When I use gdb to debug it, I got the below error.
`Starting program: /root/volume/mkl-dnn/build/examples/simple-net-cpp 
warning: Error disabling address space randomization: Operation not permitted
During startup program terminated with signal SIGSEGV, Segmentation fault.`

**Comment by nwpu2fjl** [comment]
Solved the last problem, I have encountered a new problem(seen below picture). When I build MKL-DNN with Debug. I debug the simple_net_c(segmentation fault) step by step and find that line 122 of simple_net.c causes Segmentation fualt.
![image](https://user-images.githubusercontent.com/12624852/55791361-ea3c9300-5af0-11e9-9ce4-427215af77cb.png)

**Comment by nwpu2fjl** [comment]
Finally I solve the dlopen errors(segmentation fault) by adding -DMKLDNN_ENABLE_JIT_PROFILING=0 in configure process.
```
mkdir -p build && cd build && cmake $CMAKE_OPTIONS  -DMKLDNN_THREADING=SEQ -DMKLDNN_USE_MKL=NONE -DCMAKE_BUILD_TYPE=Debug -DMKLDNN_ENABLE_JIT_PROFILING=0 ..
```

**Comment by rsdubtso** [comment]
This is weird. Which MKL-DNN version and which CMake version are these? The segfault above is due to missing `-ldl` on the `libmkldnn.so` link line.

**Comment by nwpu2fjl** [comment]
I used the latest source code pulled by https://github.com/intel/mkl-dnn. My CMake version is as below
cmake version 3.5.1.

**Comment by rsdubtso** [comment]
This has everything to do with `-ldl`. The segfault is because dynamic linker loader library that provides `dlopen()` is not linked to `libmkldnn.so`. I have confirmed that this is not reproducible on Ubuntu 18.10 on unmodified master, so probably this is due to local modifications.

I'm closing the issue and marking it as invalid. Please feel free to reopen if you observe the issue without local modifications.

**Comment by rsdubtso** [comment]
Taking that back... there is indeed a bug in CMake that causes -ldl to be ignored. Marking as a bug...

**Comment by echeresh** [comment]
@rsdubtso There might be an issue with combining `-Wl,--as-needed` (from `cmake/OpenMP.cmake`) with `-ldl` as `dlopen`, `dlclose`, etc are declared as weak symbols in `ittnotify_config.h` (is there any reason?). In this case `dl` library is discarded by the linker as there are no strong symbols from `dl` library to resolve.

**Comment by rsdubtso** [comment]
@echeresh Yes, it is exactly the problem, thanks for pointing me to `ittnotify_config,h`. I will probably need to change the definitions there...

**Comment by rsdubtso** [comment]
This is fixed in 277a32284d86f2ce4c1035d393b82c9ad9833e36 and f7a957e7ee3ed18f689a2e4cd265bf600b3bb4da
